### PR TITLE
Shop billing data fixtures

### PR DESCRIPTION
**Pull request by Prometee**
Add ability to set `invoicing_plugin_shop_billing_data` fixture

### CODE REVIEW

**Comment by Prometee** [comment]
I don't understand the error :
```
 ------ ------------------------------------------------------------------------- 
  Line   src/Fixture/ShopBillingDataFixture.php                                   
 ------ ------------------------------------------------------------------------- 
  23     Call to an undefined method                                              
         Symfony\Component\Config\Definition\Builder\NodeParentInterface::scalar  
         Node().                                                                  
  23     Calling method scalarNode() on possibly null value of type               
         Symfony\Component\Config\Definition\Builder\NodeParentInterface|null.    
 ------ ------------------------------------------------------------------------- 
 [ERROR] Found 2 errors                                                         
Script vendor/bin/phpstan.phar analyse -c phpstan.neon -l max src/ handling the analyse event returned with error code 1
``` 

How CoreBundle can get ride of this error, because they do exactly the same code as mine.

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Sylius\InvoicingPlugin\Fixture\Factory;
+
+
```
Comment (Zales0123) [inline src/Fixture/Factory/ShopBillingDataExampleFactory.php]: Doubled blank line

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
+<?php
+
```
Comment (Zales0123) [inline src/Fixture/Factory/ShopBillingDataExampleFactory.php]: Missing `declare(strict_types=1);`

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+use Sylius\Bundle\CoreBundle\Fixture\Factory\ExampleFactoryInterface;
+use Sylius\Component\Addressing\Model\CountryInterface;
+use Sylius\Component\Channel\Context\ChannelNotFoundException;
+use Sylius\Component\Channel\Repository\ChannelRepositoryInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\InvoicingPlugin\Entity\ShopBillingData;
+use Sylius\InvoicingPlugin\Entity\ShopBillingDataAwareInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class ShopBillingDataExampleFactory extends AbstractExampleFactory implements ExampleFactoryInterface
```
Comment (Zales0123) [inline src/Fixture/Factory/ShopBillingDataExampleFactory.php]: Should be `final`

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * @var OptionsResolver
+     */
+    private $optionsResolver;
+
+    /**
+     * BillingDataExampleFactory constructor.
+     * @param ChannelRepositoryInterface $channelRepository
+     */
```
Comment (Zales0123) [inline src/Fixture/Factory/ShopBillingDataExampleFactory.php]: Redundant docblock

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    public function __construct(
+        ChannelRepositoryInterface $channelRepository
```
Comment (Zales0123) [inline src/Fixture/Factory/ShopBillingDataExampleFactory.php]: Would fit in one line

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function create(array $options = []): ShopBillingDataAwareInterface
+    {
+        $options = $this->optionsResolver->resolve($options);
+
+        /** @var ShopBillingDataAwareInterface $channel */
+        $channel = $this->channelRepository->findOneByCode($options['channel']);
+        if (!$channel) {
```
Comment (Zales0123) [inline src/Fixture/Factory/ShopBillingDataExampleFactory.php]: `($channel === null)`

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    protected function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver
+            ->setRequired('channel')
+            ->setAllowedTypes('channel', ['string', ChannelInterface::class])
+            ->setRequired('company')
+            ->setAllowedTypes('company', 'string')
+            ->setRequired('country_code')
+            ->setAllowedTypes('country_code', ['string', CountryInterface::class])
```
Comment (Zales0123) [inline src/Fixture/Factory/ShopBillingDataExampleFactory.php]: I don't think we should allow `CountryInterface` here, the same for `'channel'`

File: src/Fixture/ShopBillingDataFixture.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
+<?php
+
```
Comment (Zales0123) [inline src/Fixture/ShopBillingDataFixture.php]: Missing `declare(strict_types=1);`

File: src/Fixture/ShopBillingDataFixture.php [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+namespace Sylius\InvoicingPlugin\Fixture;
+
+use Sylius\Bundle\CoreBundle\Fixture\AbstractResourceFixture;
+use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
+
+class ShopBillingDataFixture extends AbstractResourceFixture
```
Comment (Zales0123) [inline src/Fixture/ShopBillingDataFixture.php]: Should be `final`

File: src/Resources/config/services.xml [hunk 10]
```diff
@@ -112,5 +112,16 @@
...
+        <service id="sylius_invoicing_plugin.fixture.example_factory.invoicing_plugin_shop_billing_data" class="Sylius\InvoicingPlugin\Fixture\Factory\ShopBillingDataExampleFactory" public="true">
+            <argument id="sylius.repository.channel" type="service"/>
+        </service>
+        
+        <service id="sylius_invoicing_plugin.fixture.shop_billing_data_fixture" class="Sylius\InvoicingPlugin\Fixture\ShopBillingDataFixture" public="true">
+            <argument id="sylius.manager.channel" type="service"/>
+            <argument id="sylius_invoicing_plugin.fixture.example_factory.invoicing_plugin_shop_billing_data" type="service"/>
+            <tag name="sylius_fixtures.fixture"/>
+        </service>
+
```
Comment (Zales0123) [inline src/Resources/config/services.xml]: Redundant blank line

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php (hunk 7 above)
Comment (Zales0123) [reply src/Fixture/Factory/ShopBillingDataExampleFactory.php]: Yes, but this is a `channel_code`, it should never be an object :) the same for the channel, we use this value to find channel by code, so it **must** be channel code

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+final class ShopBillingDataExampleFactory extends AbstractExampleFactory implements ExampleFactoryInterface
+{
+    /**
+     * @var ChannelRepositoryInterface
+     */
+    private $channelRepository;
+
+    /**
+     * @var OptionsResolver
+     */
```
Comment (Zales0123) [inline src/Fixture/Factory/ShopBillingDataExampleFactory.php]: The convention in this plugin is to use short docblocks, so `/** @var OptionsResolver */` (same above)

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    private $optionsResolver;
+
+    public function __construct(ChannelRepositoryInterface $channelRepository) {
```
Comment (Zales0123) [inline src/Fixture/Factory/ShopBillingDataExampleFactory.php]: Move bracket to the next line

File: src/Fixture/Factory/ShopBillingDataExampleFactory.php (hunk 7 above)
Comment (Prometee) [reply src/Fixture/Factory/ShopBillingDataExampleFactory.php]: Ok get it ;)

**Comment by bartoszpietrzak1994** [comment]
Since there are 4 files changed and 15 commits, it would be nice to squash some of this commits (especially the formatting-related ones).

**Comment by Prometee** [comment]
It should be ok now ;)

File: phpstan.neon [hunk 13]
```diff
@@ -9,3 +9,6 @@ parameters:
 
     ignoreErrors:
         - '/Parameter #1 $configuration of method Symfony\Component\DependencyInjection\Extension\Extension::processConfiguration() expects Symfony\Component\Config\Definition\ConfigurationInterface, Symfony\Component\Config\Definition\ConfigurationInterface|null given./'
+        - '/Call to an undefined method Symfony\\Component\\Config\\Definition\\Builder\\NodeParentInterface/'
+        - '/Calling method scalarNode\(\) on possibly null value of type Symfony\\Component\\Config\\Definition\\Builder\\NodeParentInterface\|null/'
+        - '/Strict comparison using === between Sylius\\InvoicingPlugin\\Entity\\ShopBillingDataAwareInterface/'
```
Comment (pamil) [inline phpstan.neon]: Is this required?

**Comment by Prometee** [comment]
@pamil Yes I just need time to build it correctly and discover how you usually build those tests.
A doc should be done also.

**Comment by Prometee** [comment]
Spec tests are not possible because the class is final but I add a simple Unit test

**Comment by bartoszpietrzak1994** [comment]
After these fixes it's ok for me, WDYT @pamil @Zales0123 ?

File: tests/Unit/Fixture/ShopBillingDataFixtureTest.php [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+use Sylius\InvoicingPlugin\Fixture\ShopBillingDataFixture;
+use Sylius\Bundle\CoreBundle\Fixture\Factory\ExampleFactoryInterface;
+
+final class ShopBillingDataFixtureTest extends TestCase
+{
+    use ConfigurationTestCaseTrait;
+
+    /**
+     * @test
+     */
```
Comment (Zales0123) [inline tests/Unit/Fixture/ShopBillingDataFixtureTest.php]: Maybe we can use short docblocks? `/** @test */`

File: docs/fixtures.md [hunk 15]
```diff
@@ -0,0 +1,58 @@
+### Fixtures configuration instructions
+
+1. Add a new yaml file to the folder `config/packages` and name it `my_own_company_fixtures.yaml`
+
+2. Fill this yaml with your own company fixtures and don't forget to put first the definition of
+   your channel(s)
+   
+   You must define the channel which you will use before the definition of `invoicing_plugin_shop_billing_data`.
```
[the remark this answers is not in the document]
Comment (Prometee) [reply docs/fixtures.md]: Channel is required so I change this sentence to explain that you have to define it or use an existing one.

**Pull request merged**